### PR TITLE
Converting unoriented polygon soup to polygon mesh

### LINKED ISSUE #7647: Polygonal_surface_reconstruction produces open mesh with some inverted faces
## Issue Details

Polygonal_surface_reconstruction is producing a mesh with some face inversion, leading to a non-closed mesh.

I believe this is unexpected, since it should produce a "compact and watertight surface mesh".

I have provided an example below, which should reconstruct from the planes of a triangulated cube.

Please let me know if you see any error or misunderstanding or cannot reproduce the issue.

## Source Code

https://gist.github.com/pentacular/c02a8dea2a74006716a41c618e700d4c

#define CGAL_USE_GLPK
      
#include <CGAL/Exact_predicates_inexact_constructions_kernel.h>
#include <CGAL/GLPK_mixed_integer_program_traits.h>
#include <CGAL/Polygonal_surface_reconstruction.h>
#include <CGAL/Surface_mesh.h>
#include <CGAL/property_map.h>
      
int main() {
    typedef CGAL::Exact_predicates_inexact_constructions_kernel Epick_kernel;
    typedef Epick_kernel::Point_3 Epick_point;
    typedef Epick_kernel::Vector_3 Epick_vector;
    typedef CGAL::Surface_mesh<Epick_point> Epick_surface_mesh;
      
    typedef boost::tuple<Epick_point, Epick_vector, int> PNI;
    typedef CGAL::Nth_of_tuple_property_map<0, PNI> Point_map;
    typedef CGAL::Nth_of_tuple_property_map<1, PNI> Normal_map; 
    typedef CGAL::Nth_of_tuple_property_map<2, PNI> Plane_id_map;
    typedef std::vector<PNI> Point_vector;
      
    Point_vector points{
      PNI(Epick_point(0.5, 0.5, -1.5), Epick_vector(0, 0, -1), 0),
      PNI(Epick_point(1.5, -1.5, -1.5), Epick_vector(0, 0, -1), 0),
      PNI(Epick_point(-1.5, 1.5, -1.5), Epick_vector(0, 0, -1), 0),
      PNI(Epick_point(1.5, 1.5, -1.5), Epick_vector(0, 0, -1), 0),
      PNI(Epick_point(-0.5, -0.5, -1.5), Epick_vector(0, 0, -1), 0),
      PNI(Epick_point(-1.5, -1.5, -1.5), Epick_vector(0, 0, -1), 0),
      PNI(Epick_point(-1.5, 1.5, -1.5), Epick_vector(0, 0, -1), 0),
      PNI(Epick_point(1.5, -1.5, -1.5), Epick_vector(0, 0, -1), 0),
      PNI(Epick_point(0.5, 0.5, 1.5), Epick_vector(0, 0, 1), 1),
      PNI(Epick_point(1.5, 1.5, 1.5), Epick_vector(0, 0, 1), 1),
      PNI(Epick_point(-1.5, 1.5, 1.5), Epick_vector(0, 0, 1), 1),
      PNI(Epick_point(1.5, -1.5, 1.5), Epick_vector(0, 0, 1), 1),
      PNI(Epick_point(-0.5, -0.5, 1.5), Epick_vector(0, 0, 1), 1),
      PNI(Epick_point(1.5, -1.5, 1.5), Epick_vector(0, 0, 1), 1),
      PNI(Epick_point(-1.5, 1.5, 1.5), Epick_vector(0, 0, 1), 1),
      PNI(Epick_point(-1.5, -1.5, 1.5), Epick_vector(0, 0, 1), 1),
      PNI(Epick_point(1.5, 0.5, 0.5), Epick_vector(1, 0, 0), 2),
      PNI(Epick_point(1.5, 1.5, -1.5), Epick_vector(1, 0, 0), 2),
      PNI(Epick_point(1.5, 1.5, 1.5), Epick_vector(1, 0, 0), 2),
      PNI(Epick_point(1.5, -1.5, 1.5), Epick_vector(1, 0, 0), 2),
      PNI(Epick_point(1.5, -0.5, -0.5), Epick_vector(1, 0, 0), 2),
      PNI(Epick_point(1.5, -1.5, 1.5), Epick_vector(1, 0, 0), 2),
      PNI(Epick_point(1.5, -1.5, -1.5), Epick_vector(1, 0, 0), 2),
      PNI(Epick_point(1.5, 1.5, -1.5), Epick_vector(1, 0, 0), 2),
      PNI(Epick_point(-0.5, 1.5, 0.5), Epick_vector(0, 1, 0), 3),
      PNI(Epick_point(-1.5, 1.5, -1.5), Epick_vector(0, 1, 0), 3),
      PNI(Epick_point(-1.5, 1.5, 1.5), Epick_vector(0, 1, 0), 3),
      PNI(Epick_point(1.5, 1.5, 1.5), Epick_vector(0, 1, 0), 3),
      PNI(Epick_point(0.5, 1.5, -0.5), Epick_vector(0, 1, 0), 3),
      PNI(Epick_point(1.5, 1.5, 1.5), Epick_vector(0, 1, 0), 3),
      PNI(Epick_point(1.5, 1.5, -1.5), Epick_vector(0, 1, 0), 3),
      PNI(Epick_point(-1.5, 1.5, -1.5), Epick_vector(0, 1, 0), 3),
      PNI(Epick_point(-1.5, -0.5, 0.5), Epick_vector(-1, 0, 0), 4),
      PNI(Epick_point(-1.5, -1.5, -1.5), Epick_vector(-1, 0, 0), 4),
      PNI(Epick_point(-1.5, -1.5, 1.5), Epick_vector(-1, 0, 0), 4),
      PNI(Epick_point(-1.5, 1.5, 1.5), Epick_vector(-1, 0, 0), 4),
      PNI(Epick_point(-1.5, 0.5, -0.5), Epick_vector(-1, 0, 0), 4),
      PNI(Epick_point(-1.5, 1.5, 1.5), Epick_vector(-1, 0, 0), 4),
      PNI(Epick_point(-1.5, 1.5, -1.5), Epick_vector(-1, 0, 0), 4),
      PNI(Epick_point(-1.5, -1.5, -1.5), Epick_vector(-1, 0, 0), 4),
      PNI(Epick_point(0.5, -1.5, 0.5), Epick_vector(0, -1, 0), 5),
      PNI(Epick_point(1.5, -1.5, -1.5), Epick_vector(0, -1, 0), 5),
      PNI(Epick_point(1.5, -1.5, 1.5), Epick_vector(0, -1, 0), 5),
      PNI(Epick_point(-1.5, -1.5, 1.5), Epick_vector(0, -1, 0), 5),
      PNI(Epick_point(-0.5, -1.5, -0.5), Epick_vector(0, -1, 0), 5),
      PNI(Epick_point(-1.5, -1.5, 1.5), Epick_vector(0, -1, 0), 5),
      PNI(Epick_point(-1.5, -1.5, -1.5), Epick_vector(0, -1, 0), 5),
      PNI(Epick_point(1.5, -1.5, -1.5), Epick_vector(0, -1, 0), 5)
    };

    typedef CGAL::Polygonal_surface_reconstruction<Epick_kernel>
        Polygonal_surface_reconstruction;
    typedef CGAL::GLPK_mixed_integer_program_traits<double> MIP_Solver;

    Polygonal_surface_reconstruction algo(points, Point_map(), Normal_map(),
                                          Plane_id_map());

    Epick_surface_mesh mesh;
    if (!algo.reconstruct<MIP_Solver>(mesh)) {
      std::cerr << " Failed: " << algo.error_message() << std::endl;
      return 1;
    }
    std::cout << "Reconstructed mesh is closed: " << CGAL::is_closed(mesh) << std::endl;
    std::cout << "Reconstructed mesh: " << mesh << std::endl;
    return 0;
}

Which produces

Reconstructed mesh is closed: 0
Reconstructed mesh: OFF
24 6 0

1.5 1.5 -1.5
-1.5 1.5 -1.5
-1.5 -1.5 -1.5
1.5 -1.5 -1.5
-1.5 -1.5 1.5
1.5 1.5 1.5
-1.5 1.5 1.5
1.5 -1.5 1.5
-1.5 -1.5 1.5
1.5 1.5 -1.5
1.5 1.5 1.5
1.5 -1.5 1.5
1.5 -1.5 -1.5
-1.5 1.5 -1.5
-1.5 -1.5 -1.5
1.5 1.5 1.5
1.5 1.5 -1.5
1.5 -1.5 1.5
1.5 -1.5 -1.5
-1.5 1.5 -1.5
-1.5 1.5 1.5
-1.5 1.5 1.5
-1.5 -1.5 1.5
-1.5 -1.5 -1.5
4  2 3 0 1
4  4 7 5 6
4  11 12 9 10
4  13 21 15 16
4  14 22 17 18
4  8 23 19 20

Visualized with meshlab, we see inverted faces

Screenshot from 2023-08-11 17-45-01

## Environment

* Operating system (Linux, 64 bits):
* Compiler: g++
* Release or debug mode:
* Specific flags used (if any): -lglpk -lgmp
* CGAL version: 5.5-I-900
* Boost version:
* Other libraries versions if used (Eigen, TBB, etc.): glpk-5, gmp-6.2.1

### PR DESCRIPTION
**Pull request by soesau**
## Summary of Changes

Changed the output from unoriented polygon soup to polygon mesh

## Release Management

* Affected package(s):
* Issue(s) solved (if any): fix #7647, #7096
* Feature/Small Feature (if any):
* Link to compiled documentation (obligatory for small feature) [*wrong link name to be changed*](httpssss://wrong_URL_to_be_changed/Manual/Pkg)
* License and copyright ownership:

### CODE REVIEW

File: Polygonal_surface_reconstruction/doc/Polygonal_surface_reconstruction/dependencies [hunk 1]
```diff
@@ -5,5 +5,6 @@ Circulator
 Alpha_shapes_2
 Surface_mesh
 Point_set_processing_3
+Polygon_mesh_processing
```
Comment (sloriot) [inline Polygonal_surface_reconstruction/doc/Polygonal_surface_reconstruction/dependencies:8]: doc does not needed IIUC

**Comment by sloriot** [comment]
Test failing in CGAL-6.0-Ic-46 (segfault)

For example here:
https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-6.0-Ic-46/Polygonal_surface_reconstruction/TestReport_lrineau_Ubuntu-latest-GCC6-Release.gz

**Comment by soesau** [comment]
Can we retest this pr in the testsuite? It failed due to using PMP::orient_polygon_soup, while the orient_polygon_soup test itself also failed in the same test run.

**Comment by sloriot** [comment]
Successfully tested in CGAL-6.0-Ic-86

Issues mentioned above remain but you are right that it is a problem of the called function on those platforms.

File: Polygonal_surface_reconstruction/doc/Polygonal_surface_reconstruction/dependencies (hunk 1 above)
Comment (soesau) [reply Polygonal_surface_reconstruction/doc/Polygonal_surface_reconstruction/dependencies]: removed in next commit

**Pull request merged**